pkg/ccl/changefeedccl/event_processing.go

```diff
@@ -461,11 +468,64 @@ func (c *kvEventToRowConsumer) encodeAndEmit(
 		return err
 	}
 	if log.V(3) {
-		log.Infof(ctx, `r %s: %s -> %s`, updatedRow.TableName, keyCopy, valueCopy)
+		log.Infof(ctx, `r %s: %s(%+v) -> %s`, updatedRow.TableName, keyCopy, headers, valueCopy)
 	}
 	return nil
 }
 
+var jsonHeaderWrongTypeLogLim = log.Every(1 * time.Minute)
+var jsonHeaderWrongValTypeLogLim = log.Every(1 * time.Minute)
+
+type rowHeaders map[string][]byte
+
+func (c *kvEventToRowConsumer) makeRowHeaders(
+	ctx context.Context, updatedRow cdcevent.Row,
+) (rowHeaders, error) {
+	if c.encodingOpts.HeadersJSONColName == "" {
+		return nil, nil
+	}
+
+	headersJSON, err := readOneJSONValue(updatedRow, c.encodingOpts.HeadersJSONColName)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read headers column %s: %s", c.encodingOpts.HeadersJSONColName, updatedRow.DebugString())
+	}
+	// Allow & ignore both kinds of null.
+	if headersJSON == nil || headersJSON.Type() == json.NullJSONType {
+		return nil, nil
+	}
+
+	headers := make(rowHeaders, headersJSON.Len())
+	objIt, err := headersJSON.ObjectIter()
+	if err != nil {
+		return nil, err
+	}
+	if objIt == nil {
+		if jsonHeaderWrongTypeLogLim.ShouldLog() || log.V(2) {
+			log.Warningf(ctx, "headers column %s must be a JSON object, got %s - %s", c.encodingOpts.HeadersJSONColName, headersJSON.Type(), headersJSON)
+		}
+		return nil, nil
+	}
+	// Format the object's kv pairs. The intended use case is to have string
+	// values, but permit other types, formatted as JSON. I'd rather have
+	// confusing headers than the feed choke on specific values.
+	for objIt.Next() {
+		var val string
+		switch objIt.Value().Type() {
+		case json.StringJSONType:
```

Comment:
let's accept numbers & booleans, and skip & warn on arrays and nested objects. also skip nulls in a chill way.

Reply:
```

no warn
{"a": "b"} -> ["a", "b"]
{"a": 42} -> ["a", "42"]
{"a": true} -> ["a", "true"]

skip & warn
{"a": [1,2,3]} -> ["a", "[1, 2, 3]"]
{"a": {"b": "c"}} -> ["a", "{\"b\": \"c\"}"]
```